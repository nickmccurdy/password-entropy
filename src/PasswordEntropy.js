import CustomOptions from './CustomOptions'
import DicewareOptions from './DicewareOptions'
import GenericOptions from './GenericOptions'
import handleChange from './handleChange'
import React, { PureComponent } from 'react'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import 'react-tabs/style/react-tabs.css'

export default class PasswordEntropy extends PureComponent {
  static optionsComponents = [
    DicewareOptions,
    CustomOptions,
    GenericOptions
  ]
  static tabs = PasswordEntropy.optionsComponents.map(({ name }) => (
    <Tab key={name}>{name.replace(/Options$/, '')}</Tab>
  ))

  handleChange = handleChange.bind(this)
  handlePossibleItemsChange = possibleItems => this.setState({ possibleItems })
  state = {
    length: 6
  }
  tabPanels = PasswordEntropy.optionsComponents.map(Component => (
    <TabPanel key={Component.name}>
      <Component onChange={this.handlePossibleItemsChange}/>
    </TabPanel>
  ))

  render () {
    const possiblePasswords = (this.state.possibleItems || 1) ** this.state.length
    const approximatePrefix = possiblePasswords > Number.MAX_SAFE_INTEGER && '~ '

    return (
      <form>
        <h1>Password Entropy</h1>

        <label>
          <h2>Length</h2>
          <input name="length" value={this.state.length} onChange={this.handleChange} type="number" min="1" required/>
        </label>

        <h2>Options</h2>

        <Tabs>
          <TabList>{this.constructor.tabs}</TabList>
          {this.tabPanels}
        </Tabs>

        <h2>Possible Passwords</h2>
        {approximatePrefix}{possiblePasswords.toLocaleString()} ({approximatePrefix}{Math.log2(possiblePasswords).toFixed(2)} bits of entropy)
      </form>
    )
  }
}
