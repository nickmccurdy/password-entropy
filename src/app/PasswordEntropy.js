import CustomOptions from '../options/CustomOptions'
import DicewareOptions from '../options/DicewareOptions'
import FormGroup from '../ui/FormGroup'
import GenericOptions from '../options/GenericOptions'
import PossiblePasswords from './PossiblePasswords'
import PropTypes from 'prop-types'
import { pick } from 'ramda'
import React, { PureComponent } from 'react'
import { Form, PageHeader } from 'react-bootstrap'
import Icon from 'react-fa'
import { connect } from 'react-redux'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import 'react-tabs/style/react-tabs.css'

export class PasswordEntropy extends PureComponent {
  static options = [
    {
      Component: DicewareOptions,
      icon: 'book'
    },
    {
      Component: CustomOptions,
      icon: 'question-circle'
    },
    {
      Component: GenericOptions,
      icon: 'check-square'
    }
  ]

  static propTypes = {
    length: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
  }

  render () {
    return (
      <div>
        <PageHeader>
          <Icon name="lock"/> Password Entropy
          <br/>
          <small>A simple entropy calculator for evaluating password security.</small>
        </PageHeader>


        <Form horizontal>
          <FormGroup id="length" label="Length" icon="arrows-h">
            <input name="length" value={this.props.length} onChange={this.props.onChange} type="number" min="1" required/>
          </FormGroup>

          <h2><Icon name="cog"/> Options</h2>

          <Tabs>
            <TabList>
              {this.constructor.options.map(({ Component: { shortName }, icon }) => (
                <Tab key={shortName}><Icon name={icon}/> {shortName}</Tab>
              ))}
            </TabList>
            {this.constructor.options.map(({ Component }) => (
              <TabPanel key={Component.shortName}>
                <Component/>
              </TabPanel>
            ))}
          </Tabs>

          <FormGroup id="results" label="Results" icon="info-circle">
            <PossiblePasswords length={this.props.length}/>
          </FormGroup>
        </Form>
      </div>
    )
  }
}

export const mapStateToProps = pick(['length'])

export const mapDispatchToProps = {
  onChange: length => ({
    type: 'SET_LENGTH',
    payload: parseInt(length, 10)
  })
}

export default connect(mapStateToProps, mapDispatchToProps)(PasswordEntropy)
