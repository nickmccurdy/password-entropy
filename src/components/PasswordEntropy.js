import arrowsAltH from '@fortawesome/fontawesome-free-solid/faArrowsAltH'
import infoCircle from '@fortawesome/fontawesome-free-solid/faInfoCircle'
import lock from '@fortawesome/fontawesome-free-solid/faLock'
import Icon from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import React from 'react'
import FormControl from 'react-bootstrap/lib/FormControl'
import PageHeader from 'react-bootstrap/lib/PageHeader'
import { connect } from 'react-redux'

import FormGroup from './FormGroup'
import { selector, setLength } from '../reducers/length'
import Options from './Options'
import PossiblePasswords from './PossiblePasswords'

export function PasswordEntropy({ length, setLength }) {
  return (
    <div>
      <PageHeader>
        <Icon icon={lock} /> Password Entropy
        <br />
        <small>A simple entropy calculator for evaluating password security.</small>
      </PageHeader>

      <form>
        <FormGroup label="Results" icon={infoCircle}>
          <PossiblePasswords />
        </FormGroup>

        <FormGroup label="Length" icon={arrowsAltH}>
          <FormControl value={length} onChange={setLength} type="number" min="0" required />
        </FormGroup>

        <Options />
      </form>
    </div>
  )
}

PasswordEntropy.propTypes = {
  length: PropTypes.number.isRequired,
  setLength: PropTypes.func.isRequired
}

export default connect(selector, { setLength })(PasswordEntropy)
