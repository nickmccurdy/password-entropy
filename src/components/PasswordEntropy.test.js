import { shallow } from 'enzyme'
import React from 'react'

import PasswordEntropy from './PasswordEntropy'

test('PasswordEntropy', () => {
  const handleLength = jest.fn()
  const wrapper = shallow(
    <PasswordEntropy
      length={12}
      onCustom={() => {}}
      onGeneric={() => {}}
      onLength={handleLength}
      onOptionsKey={() => {}}
      options={{
        Custom: 0,
        Generic: {
          'Lowercase Letters': true,
          'Uppercase Letters': true,
          Numbers: true,
          Symbols: true
        }
      }}
      optionsKey="Generic"
    />
  )
  wrapper.find('TextField').simulate('change')
  expect(wrapper).toMatchSnapshot()
  expect(handleLength).toHaveBeenCalled()
})
