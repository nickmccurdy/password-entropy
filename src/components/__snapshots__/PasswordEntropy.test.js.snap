// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordEntropy 1`] = `
<React.Fragment>
  <WithStyles(Typography)
    gutterBottom={true}
    variant="headline"
  >
    Password Entropy
  </WithStyles(Typography)>
  <WithStyles(Typography)
    gutterBottom={true}
  >
    A simple entropy calculator for evaluating password security.
  </WithStyles(Typography)>
  <WithStyles(GridWrapper)
    container={true}
  >
    <WithStyles(GridWrapper)
      item={true}
      sm={6}
      xs={12}
    >
      <Card
        raised={false}
      >
        <WithStyles(CardContent)>
          <WithStyles(Typography)
            gutterBottom={true}
            variant="title"
          >
            Options
          </WithStyles(Typography)>
          <TextField
            autoFocus={true}
            label="Number of Items"
            min="0"
            onChange={[Function]}
            required={false}
            select={false}
            type="number"
            value={12}
          />
        </WithStyles(CardContent)>
        <Options
          onCustom={[Function]}
          onGeneric={[Function]}
          onOptionsKey={[Function]}
          options={
            Object {
              "Custom": 0,
              "Generic": Object {
                "Lowercase Letters": true,
                "Numbers": true,
                "Symbols": true,
                "Uppercase Letters": true,
              },
            }
          }
          optionsKey="Generic"
        />
      </Card>
    </WithStyles(GridWrapper)>
    <WithStyles(GridWrapper)
      item={true}
      sm={6}
      xs={12}
    >
      <PossiblePasswordsContainer
        length={12}
        options={
          Object {
            "Custom": 0,
            "Generic": Object {
              "Lowercase Letters": true,
              "Numbers": true,
              "Symbols": true,
              "Uppercase Letters": true,
            },
          }
        }
        optionsKey="Generic"
      />
    </WithStyles(GridWrapper)>
  </WithStyles(GridWrapper)>
</React.Fragment>
`;
